### PR TITLE
Add relative and absolute termination criterions and adjust printing …

### LINKED ISSUE #279: Naming conventions for absolute and relative tolerance parameters
I have been trying to get a consensus on the names for the absolute and relative tolerances in Base, see https://github.com/JuliaLang/julia/issues/17322 . I just noticed that `Optim.jl` seems to use `abs_tol` and `rel_tol` in some methods (`brent.jl` for example).

Is there any interest in trying to find one common naming convention for this?

### LINKED ISSUE #570: Output printing
The output printing leaves much to be desired IMO.

- It only gives the algorithm, not other important parameters (line search for instance)
- It prints the starting point and minimizer, which is only helpful for small instances (not completely true, it's sometimes helpful to get a quick look at the first component of these)
- It is very verbose
- It is not aligned and hard to parse
- The output does not follow a logical order

How about something like:

```
 * Result: converged | increasing objective | max iters
 * Minimizer:       [1.000000002262018,1.0000000045408348]
 * Minimum:         5.144946e-18
 * |x - x'|:        2.09e-10
 * |f(x) - f(x')|:  1.55e+00
 * |g(x)|:          3.36e-09
 * Algorithm: Conjugate Gradient with Hager-Zhang line search
 * Starting Point:  [0.0,0.0]
 * Iterations:      21
 * Objective Calls: 54
 * Gradient Calls:  39
```

### PR DESCRIPTION
**Pull request by pkofod**
…of the results types.

fixes #570 imo (but feel free to convince me otherwise) besides #726 
fixes #279 

```
 * Solution
    Minimizer:     [2.99999999995101,1.9999999999877076]
    Minimum:       1.034141e-19
    Converged:     true

 * Found with
    Algorithm:     Gradient Descent
    Initial Point: [0.8926024851198462,0.5761489647085432]

 * Convergence measures
    |x - x'|      = 2.68e-10 ≤ 0.0e+00 (x_abstol)
    |x - x'|/|x'| = 8.92e-11 ≤ 0.0e+00 (x_reltol)

    |f(x) - f(x')|         = 2.23e-18 ≤ 0.0e+00 (f_abstol)
    |f(x) - f(x')|/|f(x')| = 2.15e+01 ≤ 0.0e+00 (f_reltol)

    |g(x)| = 2.24e-10 ≤ 1.0e-08 (g_tol)

 * Work counters
    Iterations:    8
    f(x) calls:    30
    ∇f(x) calls:   30
```

for now, the old `x_tol`, `f_tol` and `g_tol` overwrite abstol, reltol and abstol respectively.

### CODE REVIEW

**Comment by antoine-levitt** [comment]
Looks good! A couple of suggestions (mostly personal preference, feel free to discard):

- Change \le to \nleq when the condition is not satisfied
- Align the numbers in the convergence measures
- Don't skip lines inside the convergence measures
- Remove the `(x_abstol)` and friends
- Add a "successful" or "unsuccessful" flag at the top, and remove the "converged: true"
- Move the "stopped by increasing function objective" into the convergence flag at the top
- Print less significant digits in the minimizers and initial point

**Comment by pkofod** [comment]
Excellent input, thanks!

>     * Change \le to \nleq when the condition is not satisfied

Good idea!

>     * Align the numbers in the convergence measures
> 
>     * Don't skip lines inside the convergence measures

I'll try it, I'm just worried the distance from |g| to the number will be too large.
 
>     * Remove the `(x_abstol)` and friends

:+1:

>     * Add a "successful" or "unsuccessful" flag at the top, and remove the "converged: true"

Yes, that's probably better.

>     * Move the "stopped by increasing function objective" into the convergence flag at the top

I actually moved it in with the "iteration limit exceeded" line just now, that is only shown upon "unsuccessful termination". 
 
>     * Print less significant digits in the minimizers and initial point

Good point!

**Comment by pkofod** [comment]
```
 * Status: failure

 * Candidate solution
    Minimizer: [3.79e-01, 5.37e-01]
    Minimum:   5.049244e-02

 * Found with
    Algorithm:     Accelerated Gradient Descent
    Initial Point: [3.79e-01, 5.37e-01]

 * Convergence measures
    |x - x'|               = 5.52e-02 ≰ 0.0e+00
    |x - x'|/|x'|          = 1.82e-02 ≰ 0.0e+00
    |f(x) - f(x')|         = 1.69e-02 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 2.51e-01 ≰ 0.0e+00
    |g(x)|                 = 1.63e+00 ≰ 1.0e-08
    Reached Maximum Number of Iterations: false
    Stopped by an increasing objective: true

 * Work counters
    Iterations:    5
    f(x) calls:    25
    ∇f(x) calls:   25
```

and

```
 * Status: success

 * Candidate solution
    Minimizer: [7.72e-01, 3.10e-02]
    Minimum:   2.317441e-18

 * Found with
    Algorithm:     Momentum Gradient Descent
    Initial Point: [7.72e-01, 3.10e-02]

 * Convergence measures
    |x - x'|               = 3.88e-10 ≰ 0.0e+00
    |x - x'|/|x'|          = 1.29e-10 ≰ 0.0e+00
    |f(x) - f(x')|         = 3.30e-18 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 1.42e+00 ≰ 0.0e+00
    |g(x)|                 = 8.57e-09 ≤ 1.0e-08

 * Work counters
    Iterations:    29
    f(x) calls:    111
    ∇f(x) calls:   111
```

**Comment by antoine-levitt** [comment]
Very nice! The only suggestion I have would be in the failure case, to not print max iter / increasing objectives as true/false, but to adjust the status message, ie

```
 * Status: failure (stopped by increasing objective)

 * Candidate solution
    Minimizer: [3.79e-01, 5.37e-01]
    Minimum:   5.049244e-02

 * Found with
    Algorithm:     Accelerated Gradient Descent
    Initial Point: [3.79e-01, 5.37e-01]

 * Convergence measures
    |x - x'|               = 5.52e-02 ≰ 0.0e+00
    |x - x'|/|x'|          = 1.82e-02 ≰ 0.0e+00
    |f(x) - f(x')|         = 1.69e-02 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 2.51e-01 ≰ 0.0e+00
    |g(x)|                 = 1.63e+00 ≰ 1.0e-08

 * Work counters
    Iterations:    5
    f(x) calls:    25
    ∇f(x) calls:   25
```

**Comment by pkofod** [comment]
for some reason, some tests fail now.. at least it gives us a chance to see the output in action :)

```
Polynomial did not converge with i = 1
 * Status: failure (reached maximum number of iterations)
 * Candidate solution
    Minimizer: [1.00e+01, 7.00e+00, 1.08e+02]
    Minimum:   6.727258e-12
 * Found with
    Algorithm:     Accelerated Gradient Descent
    Initial Point: [0.00e+00, 0.00e+00, 0.00e+00]
 * Convergence measures
    |x - x'|               = 3.20e-06 ≰ 0.0e+00
    |x - x'|/|x'|          = 2.96e-08 ≰ 0.0e+00
    |f(x) - f(x')|         = 2.58e-15 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 3.84e-04 ≰ 0.0e+00
    |g(x)|                 = 2.89e-08 ≰ 1.0e-08
 * Work counters
    Iterations:    1500
    f(x) calls:    4968
    ∇f(x) calls:   4968
```
```
 [7] top-level scope at /home/travis/build/JuliaNLSolvers/Optim.jl/test/multivariate/solvers/second_order/newton.jl:2
Paraboloid Random Matrix did not converge with i = 1
 * Status: failure (objective increased between iterations)
 * Candidate solution
    Minimizer: [1.00e+00, 1.00e+00, 1.00e+00,  ...]
    Minimum:   2.364721e-18
 * Found with
    Algorithm:     Newton's Method
    Initial Point: [0.00e+00, 0.00e+00, 0.00e+00,  ...]
 * Convergence measures
    |x - x'|               = 1.08e-10 ≰ 0.0e+00
    |x - x'|/|x'|          = 1.08e-10 ≰ 0.0e+00
    |f(x) - f(x')|         = 1.93e-24 ≰ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 8.14e-07 ≰ 0.0e+00
    |g(x)|                 = 1.39e-08 ≰ 1.0e-08
 * Work counters
    Iterations:    65
    f(x) calls:    215
    ∇f(x) calls:   215
    ∇²f(x) calls:  66
```

**Comment by pkofod** [comment]
Test failures come from DiffEqDiffTools.
